### ISSUE TITLE
Why dehydration resets dispatcher?

### ISSUE THREAD

**Issue by vslinko** [gaearon/redux#51]
```js
export default async function renderPage(url) {
  const dispatcher = createDispatcher(composeStores(stores))

  const {title} = await dispatcher.perform(runTransition(url))

  // I need to get atom to make some checks
  const {atom} = dispatcher.dehydrate()

  if (!atom.routerStore.componentKey) {
    return {
      status: 404
    }
  }

  // But after that checks I should return atom into dispatcher for render usages
  // I want to remove this step
  dispatcher.hydrate({atom})

  const pageContent = React.renderToString(
    <Provider dispatcher={dispatcher}>
      {() => <Application dispatcher={dispatcher} />}
    </Provider>
  )

  return {
    body: appTemplate(title, pageContent, atom)
  }
}
```

**Comment by gaearon** [comment]
Should I add a separate method like `dispose`? I'm not sure what the convention is regarding (de)hydration and disposal.

**Comment by acdlite** [comment]
It seems we just need a way to access the atom.

`dispose()` seems like the better word for a method that clears the internal state. `dehydrate()` in other Flux libraries more closely means "convert state to a serialized string so I have send it to the client."

**Comment by vslinko** [comment]
I can't imagine that someone needs to dispose dispatcher.
Btw I have some API proposes:
```
// 1. current
const hydro = dispatcher.dehydrate()
dispatcher.hydrate(hydro)

// 2. extract dispose
const hydro = dispatcher.dehydrate()
dispatcher.dispose()
dispatcher.hydrate(hydro)

// 3. create getAtom method
const atom = dispatcher.getAtom()
dispatcher.hydrate({atom})
```

**Comment by acdlite** [comment]
We need to dispose the dispatcher inside `componentWillUpdate()` to support prop changes / hot reloading. Right now we use the methods `dehydrate()` and `hydrate()` but I propose that `dispose()` and `initialize()` are closer to what those methods actually do.

**Comment by acdlite** [comment]
:+1: for `getAtom()` method

**Comment by gaearon** [comment]
Yes, the use case for disposing the dispatcher is hot reloading.

I'm done for today but I'd merge a PR that addresses this. @acdlite Can you come up with an API to satisfy both use cases? I don't mind adding more methods if it makes server rendering more convenient.

**Comment by acdlite** [comment]
For server rendering:
- `getAtom(atom): atom`
- `setAtom(atom)`

For hot reloading:
- `initialize(atom, subscriptions)`
- `dispose(): { atom, subsciptions }`

**Comment by gaearon** [comment]
Looks good to me!

Maybe initialize({ atom, subscriptions })? So it matches the shape.

**Comment by acdlite** [comment]
Good idea!

**Comment by gaearon** [comment]
I just added `getAtom` in master (for a different purpose though): ff56611fdb3b599519fd8998dcd8d650d1e0b683

**Comment by gaearon** [comment]
Closed by #54
